### DISCUSSION TITLE
『浅入深出』MySQL 中事务的实现 · /mysql-transaction

### DISCUSSION THREAD

**Discussion by draveness** [draveness/blog-comments#252] [Announcements]
https://draveness.me/mysql-transaction

**Comment by asfan** [comment]
理论清晰清楚，受教了~

**Comment by draveness** [comment]
> @asfan
> 理论清晰清楚，受教了~

感谢支持

**Comment by hungryTechBoy** [comment]
非常受用，感谢分享

**Comment by huanghantao** [comment]
👏🏻

**Comment by darkerzhang** [comment]
@Draveness 请问 你整理出这一篇文章大概需要多长时间  一天？

**Comment by draveness** [comment]
@darkerzhang 你想多了。。一周吧

**Comment by ihaohong** [comment]
讲解深入浅出，画的图也非常有逻辑性~ 请问你平常画图都是用的什么工具？

**Comment by Sonion** [comment]
最好能够增加参考文献，帮助一起理解

**Comment by draveness** [comment]
> @Sonion
> 最好能够增加参考文献，帮助一起理解

这不是有 reference 么

**Comment by draveness** [comment]
> @Choaya-Elvis
> 写得很好，我想把这篇文章引入我的简书博客中，征求作者意见。

转载注明链接

**Comment by gongxk27** [comment]
你好，请问你的作图工具是什么啊

**Comment by yelcat** [comment]
我也想问：请问你用的是什么画图软件

**Comment by iobee** [comment]
清晰明了。赞

**Comment by yuanjize** [comment]
“在上图所示的一次事件中，Transaction2 依赖于 Transaction1，而 Transaction3 又依赖于 Transaction1，当 Transaction1 由于执行出现问题发生回滚时，为了保证事务的原子性”--------这个地方是不是2依赖1,3依赖2.

**Comment by Slaughter96** [comment]
不可重复读和幻读的描述是不是反了？

**Comment by tubalu** [comment]
求Sketch 模板

**Comment by hoho1221** [comment]
请教一下，重做日志与回滚日志，针对的事务操作实现方式不一样，前者针对数据库操作提交后写入数据库，后者针对数据库操作实时写入数据库，那么两者如何能够一起工作？同时存在是否能起到其各自的作用呢？ thx~

**Comment by draveness** [comment]
> @MrCQ
> 请教一下，重做日志与回滚日志，针对的事务操作实现方式不一样，前者针对数据库操作提交后写入数据库，后者针对数据库操作实时写入数据库，那么两者如何能够一起工作？同时存在是否能起到其各自的作用呢？ thx~

你对这两个日志的理解还有点问题，重做日志解决的是崩溃恢复时已提交的事务不会丢失，回滚日志解决的就是事务的回滚问题

**Comment by hoho1221** [comment]
> @Draveness
> > @MrCQ
> > 请教一下，重做日志与回滚日志，针对的事务操作实现方式不一样，前者针对数据库操作提交后写入数据库，后者针对数据库操作实时写入数据库，那么两者如何能够一起工作？同时存在是否能起到其各自的作用呢？ thx~
> 
> 你对这两个日志的理解还有点问题，重做日志解决的是崩溃恢复时已提交的事务不会丢失，回滚日志解决的就是事务的回滚问题

阅读了参考文献中关于Undo与Redo的文章后，比较疑惑

重做日志解决的是崩溃恢复时已提交的事务不会丢失，那么也就意味着数据是在事务提交之后才会被写入数据库，事务提交到写入数据库之前发生崩溃，需要通过Redo避免数据丢失
回滚日志解决的就是事务的回滚问题，事务提交之前数据已经被写入到数据库，这种情况下出现异常才会需要Undo来回滚

那么两种日志针对数据写入数据库的时机各异的实现方式，如何协同在一起工作？ 这是我现在比较疑惑的点，可能是哪个地方理解上出现偏差，thx~

**Comment by xiaofeng-ling** [comment]
非常不错，其中关于回滚日志和重做日志的理解让人豁然开朗

**Comment by andyRon** [comment]
看了博主好几遍文章，图片风格很统一，都是sketch做的？！ 看上去很舒服啊！最近也刚接触sketch。

**Comment by aiyueqi** [comment]
非常清晰，学习了，感谢！！

**Comment by asd1245dss** [comment]
第六行有错别字  我们也就~清除~了事务的实现 => 清楚
第三十七行 由于向第三方请求并获取结果是一个需要较长~事件~的操作 => 时间
时间戳段落第二行 读时间戳中~报错~了所有访问该记录的事务中的最大时间戳 => 保存

**Comment by Pramy** [comment]
图画的很清晰很漂亮，请问是用什么工具做得呢

**Comment by draveness** [comment]
> @asd1245dss
> 第六行有错别字  我们也就~清除~了事务的实现 => 清楚
> 第三十七行 由于向第三方请求并获取结果是一个需要较长~事件~的操作 => 时间
> 时间戳段落第二行 读时间戳中~报错~了所有访问该记录的事务中的最大时间戳 => 保存

改好了，多谢

**Comment by lemon-clown** [comment]
很漂亮的博客，感谢博主的分享；
另外，请问博主图片是用什么做的呢？

**Comment by ljw3713** [comment]
1.重做日志的图“Rego logging”
2.“读时间戳中<报错>了所有访问该记录的事务中的最大时间戳” 错别字

**Comment by draveness** [comment]
> @ljw8947
> 1.重做日志的图“Rego logging”
> 2.“读时间戳中<报错>了所有访问该记录的事务中的最大时间戳” 错别字
> 

修复了

**Comment by blue-troy** [comment]
牛逼

**Comment by jinhailang** [comment]
非常棒

**Comment by wxainn** [comment]
棒

**Comment by lintiyan** [comment]
有个地方没明白。事务提交时会先将重做日志缓存写入重做日志文件中，再将内存中的数据更新到磁盘。如果提交时数据库服务挂了，那么下次重启时，会根据重做日志恢复数据。所以这个重做日志是在数据库服务挂掉时的事务持久性保障。问题是，假如事务提交时，重做日志缓存还每来的及写入重做日志文件中，那么怎么保证事务的持久性，或者说会不会有重做日志缓存没有写入文件这种情况发生？

**Comment by draveness** [comment]
只有写入重做日志，才认为这是一次成功写入了，否则都认为是写入失败

**Comment by imageslr** [comment]
一口气看了作者的好几篇文章，从分布式开始一篇一篇向下读，写的真的好

**Comment by unique-cjx** [comment]
好文，顶

**Comment by ansionfor** [comment]
“所有事务进行的修改都会先记录到这个回滚日志中”，在我印象中undo log记录的是修改前的记录值，但博主这句话好像没有表达出这个意思

**Comment by XinruZhang** [comment]
> @lintiyan
> 有个地方没明白。事务提交时会先将重做日志缓存写入重做日志文件中，再将内存中的数据更新到磁盘。如果提交时数据库服务挂了，那么下次重启时，会根据重做日志恢复数据。所以这个重做日志是在数据库服务挂掉时的事务持久性保障。问题是，假如事务提交时，重做日志缓存还每来的及写入重做日志文件中，那么怎么保证事务的持久性，或者说会不会有重做日志缓存没有写入文件这种情况发生？

https://zhuanlan.zhihu.com/p/35574452 

我觉得这里解释的挺好的，重点是 Redo Log 和 Undo Log 中记录的都是什么。

首先约定，无论是 Redo 还是 Undo：
```
1. 事务开始时，记录START T
2. 事务修改时，记录（T，x，v），说明事务T操作对象x，x的值为v 
3. 事务结束时，记录COMMIT T 
```

不同点在于记录时 `(T, x, v)` 中的 `v` 的不同，根据 Undo 和 Redo 的目的：
- **Undo** ：将所有没有提交的事务恢复到提交前的状态，因而 `(T, x, v)` 中 v 记录的是修改前的值，便于恢复。
- **Redo** ：将所有已提交的事务（已写入 log，但因为崩溃数据未写入）重新写入，因而 `(T, x, v)` 中 v 记录的是修改后的值，便于重做。

**Comment by XinruZhang** [comment]
另外，很感谢博主的总结和分享，看了一下博主的其他文章，内容都很优质，值得推荐

**Comment by speedfirst** [comment]
图的配色很赞啊。请问用什么工具绘制的？

**Comment by draveness** [comment]
> @speedfirst
> 图的配色很赞啊。请问用什么工具绘制的？

不看一下文章结尾么？

**Comment by CNYuYang** [comment]
看了一下午的资料，这是我读到的最棒的文章。多谢

**Comment by oldChen89** [comment]
网站的布局是自己写的麽？这种样式大爱，有模板麽老铁？

**Comment by jiahuat** [comment]
“最后，也是最重要的，当别人在将一致性的时候”中 “在将” 是“在讲”吧？

---

*2020-06-04 UDPATES：已修复*

**Comment by draveness** [comment]
> 网站的布局是自己写的麽？这种样式大爱，有模板麽老铁？

主题是 "hugo-lithium"，自己改了一些

**Comment by tra4less** [comment]
> 而写锁保证了在更新数据库数据时不会有其他的事务访问或者更改同一条记录造成不可预知的问题


写锁并不能保证更新数据库数据时不会有其他的事务访问同一条记录。

**Comment by EeeMt** [comment]
写的挺好

**Comment by enginnerFrankLiu** [comment]
数据对并发的控制有两种
1.一种是悲观并发控制
  悲观并发控制的本质是锁
  不同事务隔离级别（snapshot以上）影响共享锁(SELECT)的生命周期
  和锁的粒度(范围)

2.另一种是乐观并发控制
  基于版本号，或者说是基于snapshot 和 当前事务中的数据的对比，对比是否有被更新过，在当前事务中

**Comment by enginnerFrankLiu** [comment]
阅读了很多文章后，个人真的不太喜欢描述“乐观并发控制”或者“乐观并发机制” 为 “乐观锁”，因为我们常见的锁，只有，共享锁，更新锁，排他锁（当然还有意向xx锁等)，根本就没有所谓的“乐观锁”， 不知道作者 怎么看？

**Comment by enginnerFrankLiu** [comment]
对 ACID 和 CAP 中一致性的讲解非常到位，赞~

**Comment by ydzh** [comment]
> @hoho1221
> > @Draveness
> > > @MrCQ
> > > 请教一下，重做日志与回滚日志，针对的事务操作实现方式不一样，前者针对数据库操作提交后写入数据库，后者针对数据库操作实时写入数据库，那么两者如何能够一起工作？同时存在是否能起到其各自的作用呢？ thx~
> > 
> > 你对这两个日志的理解还有点问题，重做日志解决的是崩溃恢复时已提交的事务不会丢失，回滚日志解决的就是事务的回滚问题
> 
> 阅读了参考文献中关于Undo与Redo的文章后，比较疑惑
> 
> 重做日志解决的是崩溃恢复时已提交的事务不会丢失，那么也就意味着数据是在事务提交之后才会被写入数据库，事务提交到写入数据库之前发生崩溃，需要通过Redo避免数据丢失
> 回滚日志解决的就是事务的回滚问题，事务提交之前数据已经被写入到数据库，这种情况下出现异常才会需要Undo来回滚
> 
> 那么两种日志针对数据写入数据库的时机各异的实现方式，如何协同在一起工作？ 这是我现在比较疑惑的点，可能是哪个地方理解上出现偏差，thx~
> 

同感，undo log应该是回滚和repeated read使用的，感觉对持久性没有帮助

**Comment by ghost** [comment]
隔离级别的标准提案在mysql的背景下那一段写得不太对，RC在mysql可以解决脏读，RR在mysql的背景下可以解决幻读

**Reply by wanghengg** [reply]
INNODB 的 RR在快照读和当前读时分别通过 MVCC 和next-key lock 能部分解决幻读的问题，但是不能完全解决。幻读仍然可能发生

**Comment by wanghengg** [comment]
关于 READ COMMITTED（读提交）的实现方式描述有问题，READ COMMITTED 和 READ REPEATED 都是通过READ VIEW 实现的，只是 READ COMMITTED 在每条 SQL 语句执行时生成 READ VIEW，但是 READ REPEATED在每个事务启动时生成 READ VIEW。
